We should detect if sudo non-functional and don't use it for rsync scripts

Got this errors in logs:
<details>

```
2017-03-27 01:00:01,933[pScheduler-6216]  [ERROR] [DockerEnvironmentBackupManager 509]  - Error occurred during backup/restore of workspace 'workspace*******' on node 'node12.codenvy.io' : Warning: Permanently added '[node12.code
Authenticated to node12.codenvy.io ([172.30.12.112]:36526).
Transferred: sent 2336, received 2220 bytes, in 0.0 seconds
Bytes per second: sent 230170.4, received 218740.7
Warning: Permanently added '[node12.codenvy.io]:36526,[172.30.12.112]:36526' (ECDSA) to the list of known hosts.
Authenticated to node12.codenvy.io ([172.30.12.112]:36526).
sudo: no tty present and no askpass program specified
Transferred: sent 2396, received 2260 bytes, in 0.0 seconds
Bytes per second: sent 188812.6, received 178095.4
rsync: connection unexpectedly closed (0 bytes received so far) [Receiver]
rsync error: error in rsync protocol data stream (code 12) at io.c(226) [Receiver=3.1.2]
2017-03-27 01:00:01,935[pScheduler-6216]  [ERROR] [DockerEnvironmentBackupManager 356]  - Synchronization process failed. Exit code 12
java.io.IOException: Synchronization process failed. Exit code 12
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeCommand(DockerEnvironmentBackupManager.java:511) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeBackupScript(DockerEnvironmentBackupManager.java:347) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupInsideLock(DockerEnvironmentBackupManager.java:228) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupWorkspace(DockerEnvironmentBackupManager.java:127) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.WorkspaceFsBackupScheduler.lambda$scheduleBackup$0(WorkspaceFsBackupScheduler.java:106) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
2017-03-27 01:00:01,936[pScheduler-6216]  [ERROR] [m.b.WorkspaceFsBackupScheduler 112]  - Backup of workspace workspace************** filesystem terminated on node12.codenvy.io node. Synchronization process failed. Exit code 12
org.eclipse.che.api.core.ServerException: Backup of workspace workspace********** filesystem terminated on node12.codenvy.io node. Synchronization process failed. Exit code 12
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeBackupScript(DockerEnvironmentBackupManager.java:359) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupInsideLock(DockerEnvironmentBackupManager.java:228) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupWorkspace(DockerEnvironmentBackupManager.java:127) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.WorkspaceFsBackupScheduler.lambda$scheduleBackup$0(WorkspaceFsBackupScheduler.java:106) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
```
</details>

It looks like `rsync` doesn't like extra output that `sudo` produces, resulting in error code 12.
We need to fix it, either by not only checking that `sudo` command is present but testing if it functional, or by allocating pseudo-terminal by `ssh -t` if we know user password or by asking user for permission to modify sudo configuration to allow passwordless use of sudo.

Maybe it worth to consider using `dropbear` and/or `gosudo` instead of distributive specific versions, that may have altered configuration in images/recipes users trying to use.

**Reproduction Steps:**
Create new workspace from stack, I used python console simple.
Run workspace, open terminal window, and as root edit `/etc/sudoers` file:
just comment out last line that allows passwordless use of sudo:
```
#includedir /etc/sudoers.d
#%sudo ALL=(ALL) NOPASSWD: ALL
```
wait some time like 10 minutes and observe errors in log:
<details>

```
2017-03-28 13:30:38,956[kupScheduler-68]  [ERROR] [DockerEnvironmentBackupManager 509]  - Error occurred during backup/restore of workspace 'workspacebebdz483vl2b1hs9' on node 'node2.c5.codenvy-stg.com' : Warning: Permanently added '[node
2.c5.codenvy-stg.com]:40510,[172.30.12.84]:40510' (ECDSA) to the list of known hosts.
Warning: Permanently added '[node2.c5.codenvy-stg.com]:40510,[172.30.12.84]:40510' (ECDSA) to the list of known hosts.
sudo: no tty present and no askpass program specified
rsync: connection unexpectedly closed (0 bytes received so far) [Receiver]
rsync error: error in rsync protocol data stream (code 12) at io.c(226) [Receiver=3.1.2]
2017-03-28 13:30:38,957[kupScheduler-68]  [ERROR] [DockerEnvironmentBackupManager 356]  - Synchronization process failed. Exit code 12
java.io.IOException: Synchronization process failed. Exit code 12
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeCommand(DockerEnvironmentBackupManager.java:511) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeBackupScript(DockerEnvironmentBackupManager.java:347) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupInsideLock(DockerEnvironmentBackupManager.java:228) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupWorkspace(DockerEnvironmentBackupManager.java:127) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.WorkspaceFsBackupScheduler.lambda$scheduleBackup$0(WorkspaceFsBackupScheduler.java:106) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
2017-03-28 13:30:38,958[kupScheduler-68]  [ERROR] [m.b.WorkspaceFsBackupScheduler 112]  - Backup of workspace workspacebebdz483vl2b1hs9 filesystem terminated on node2.c5.codenvy-stg.com node. Synchronization process failed. Exit code 12
org.eclipse.che.api.core.ServerException: Backup of workspace workspacebebdz483vl2b1hs9 filesystem terminated on node2.c5.codenvy-stg.com node. Synchronization process failed. Exit code 12
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeBackupScript(DockerEnvironmentBackupManager.java:359) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupInsideLock(DockerEnvironmentBackupManager.java:228) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupWorkspace(DockerEnvironmentBackupManager.java:127) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.WorkspaceFsBackupScheduler.lambda$scheduleBackup$0(WorkspaceFsBackupScheduler.java:106) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
```
</details>

**Codenvy version:**    5.5.1-SaaS

--- a/dockerfiles/init/modules/codenvy/manifests/init.pp
+++ b/dockerfiles/init/modules/codenvy/manifests/init.pp
@@ -124,6 +124,14 @@ class codenvy {
     require => File[$config_dirs],
   }
 
+# creating rsyncbase.sh
+  file { "/opt/codenvy/config/codenvy/conf/rsyncbase.sh":
+    ensure  => "present",
+    content => template("codenvy/rsyncbase.sh.erb"),
+    mode    => "755",
+    require => File[$config_dirs],
+  }
+
 # creating rsyncbackup.sh
   file { "/opt/codenvy/config/codenvy/conf/rsyncbackup.sh":
     ensure  => "present",

--- a/dockerfiles/init/modules/codenvy/templates/rsyncbackup.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncbackup.sh.erb
@@ -1,5 +1,16 @@
 #!/bin/bash
 
+#
+# Copyright (c) 2012-2017 Codenvy, S.A.
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#   Codenvy, S.A. - initial API and implementation
+#
+
 SRC_FOLDER=${1}
 HOST=${2}
 PORT=${3}

--- a/dockerfiles/init/modules/codenvy/templates/rsyncbackup.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncbackup.sh.erb
@@ -1,49 +1,18 @@
 #!/bin/bash
 
 SRC_FOLDER=${1}
-SRC_HOST=${2}
-SRC_PORT=${3}
+HOST=${2}
+PORT=${3}
 DST_FOLDER=${4}
 REMOVE_ON_SUCCESS=${5=false}
-SRC_USER_NAME=${6}
-
-# Key for SSH connection to sync files
-SSH_KEY=/opt/codenvy-data/conf/ssh/key.pem
-# When Codenvy runs on windows we have issue since key doesn't have permissions 0600.
-# With another permissions on key SSH doesn't work.
-# So we copy key into another place that should be inside of container - so it is Linux and permissions are supported.
-# And apply needed permission to copied file. Copying is performed lazily, so it is skipped when it was done previously.
-# ensure ssh key has correct permissions
-if [[ ! -f /opt/rsync_key ]]; then
-    cp $SSH_KEY /opt/rsync_key
-fi
-if [[ $(stat -c %a /opt/rsync_key) != 600 ]]; then
-    chmod 0600 /opt/rsync_key
-fi
+USER_NAME=${6}
 
 ##### Set by puppet #####
 RSYNC_BACKUP_BWLIMIT=<%= scope.lookupvar('codenvy::rsync_backup_bwlimit') %>
 SSH_LOG_LEVEL=<%= scope.lookupvar('codenvy::rsync_ssh_log_level') %>
 #########################
 
-##### SSH options #####
-SSH_OPTIONS=""
-# Add SSH connection options
-SSH_OPTIONS+=" -i /opt/rsync_key -l ${SRC_USER_NAME} -p ${SRC_PORT}"
-# Disable password authentication since we use key-based auth
-SSH_OPTIONS+=" -o PasswordAuthentication=no"
-# Disable hosts fingerprint checking because it may fail due to
-# starting different containers on the same ports after some period
-SSH_OPTIONS+=" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
-# Set SSH logging level to make it possible to investigate problems
-SSH_OPTIONS+=" -o LogLevel=${SSH_LOG_LEVEL}"
-#######################
-
-# We need root permissions in target container to change ownership of files.
-# If user in that container has access to sudo we use it.
-# Otherwise we use rsync without sudo and if it is possible ownership will be changed.
-RSYNC_COMMAND=$(ssh ${SSH_OPTIONS} ${SRC_HOST} \
-            "if hash sudo 2>/dev/null; then echo 'sudo rsync'; else echo 'rsync'; fi")
+source /opt/codenvy-data/conf/rsyncbase.sh
 
 ##### Rsync options #####
 RSYNC_OPTIONS=""
@@ -80,4 +49,4 @@ rsync ${RSYNC_OPTIONS} \
       --rsh="ssh ${SSH_OPTIONS}" \
       --rsync-path="${RSYNC_COMMAND}" \
       --include='.codenvy/' --filter=':- .gitignore' \
-      ${SRC_HOST}:${SRC_FOLDER} ${DST_FOLDER}
+      ${HOST}:${SRC_FOLDER} ${DST_FOLDER}

--- a/dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb
@@ -1,0 +1,37 @@
+#!/bin/bash
+
+# Key for SSH connection to sync files
+SSH_KEY=/opt/codenvy-data/conf/ssh/key.pem
+# When Codenvy runs on windows we have issue since key doesn't have permissions 0600.
+# With another permissions on key SSH doesn't work.
+# So we copy key into another place that should be inside of container - so it is Linux and permissions are supported.
+# And apply needed permission to copied file. Copying is performed lazily, so it is skipped when it was done previously.
+# ensure ssh key has correct permissions
+if [[ ! -f /opt/rsync_key ]]; then
+    cp $SSH_KEY /opt/rsync_key
+fi
+if [[ $(stat -c %a /opt/rsync_key) != 600 ]]; then
+    chmod 0600 /opt/rsync_key
+fi
+
+##### SSH options #####
+SSH_OPTIONS=""
+# Add SSH connection options
+SSH_OPTIONS+=" -i /opt/rsync_key -l ${USER_NAME} -p ${PORT}"
+# Disable password authentication since we use key-based auth
+SSH_OPTIONS+=" -o PasswordAuthentication=no"
+# Disable hosts fingerprint checking because it may fail due to
+# starting different containers on the same ports after some period
+SSH_OPTIONS+=" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
+# Set SSH logging level to make it possible to investigate problems
+SSH_OPTIONS+=" -o LogLevel=${SSH_LOG_LEVEL}"
+#######################
+
+# We need root permissions in target container to change ownership of files.
+# If user in that container has access to sudo we use it.
+# Otherwise we use rsync without sudo and if it is possible ownership will be changed.
+checkFile="/tmp/codenvy_sudo_test_$(date +%s)"
+createCheckFile="sh -c 'echo 1 > $checkFile' 2>/dev/null"
+checkSudo="$createCheckFile; hash sudo 2>/dev/null && sudo chown 1000 $checkFile 2>/dev/null"
+GET_RSYNC_COMMAND='if [ "$(id -u)" != "0" ]; then '$checkSudo' && echo "sudo rsync" && exit; fi; echo "rsync"'
+RSYNC_COMMAND=$(ssh ${SSH_OPTIONS} ${HOST} $(echo "$GET_RSYNC_COMMAND"))

--- a/dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb
@@ -1,5 +1,16 @@
 #!/bin/bash
 
+#
+# Copyright (c) 2012-2017 Codenvy, S.A.
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#   Codenvy, S.A. - initial API and implementation
+#
+
 # Key for SSH connection to sync files
 SSH_KEY=/opt/codenvy-data/conf/ssh/key.pem
 # When Codenvy runs on windows we have issue since key doesn't have permissions 0600.

--- a/dockerfiles/init/modules/codenvy/templates/rsyncrestore.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncrestore.sh.erb
@@ -1,5 +1,16 @@
 #!/bin/bash
 
+#
+# Copyright (c) 2012-2017 Codenvy, S.A.
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#   Codenvy, S.A. - initial API and implementation
+#
+
 SRC_FOLDER=${1}
 DST_FOLDER=${2}
 HOST=${3}

--- a/dockerfiles/init/modules/codenvy/templates/rsyncrestore.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncrestore.sh.erb
@@ -2,48 +2,18 @@
 
 SRC_FOLDER=${1}
 DST_FOLDER=${2}
-DST_HOST=${3}
-DST_PORT=${4}
+HOST=${3}
+PORT=${4}
 DST_USR_ID=${5}
 DST_GRP_ID=${6}
-SRC_USER_NAME=${7}
-
-# Key for SSH connection to sync files
-SSH_KEY=/opt/codenvy-data/conf/ssh/key.pem
-# When Codenvy runs on windows we have issue since key doesn't have permissions 0600.
-# With another permissions on key SSH doesn't work.
-# So we copy key into another place that should be inside of container - so it is Linux and permissions are supported.
-# And apply needed permission to copied file. Copying is performed lazily, so it is skipped when it was done previously.
-if [[ ! -f /opt/rsync_key ]]; then
-    cp $SSH_KEY /opt/rsync_key
-fi
-if [[ $(stat -c %a /opt/rsync_key) != 600 ]]; then
-    chmod 0600 /opt/rsync_key
-fi
+USER_NAME=${7}
 
 ##### Set by puppet #####
 RSYNC_RESTORE_BWLIMIT=<%= scope.lookupvar('codenvy::rsync_restore_bwlimit') %>
 SSH_LOG_LEVEL=<%= scope.lookupvar('codenvy::rsync_ssh_log_level') %>
 #########################
 
-##### SSH options #####
-SSH_OPTIONS=""
-# Add SSH connection options
-SSH_OPTIONS+=" -i /opt/rsync_key -l ${SRC_USER_NAME} -p ${DST_PORT}"
-# Disable password authentication since we use key-based auth
-SSH_OPTIONS+=" -o PasswordAuthentication=no"
-# Disable hosts fingerprint checking because it may fail due to
-# starting different containers on the same ports after some period
-SSH_OPTIONS+=" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
-# Set SSH logging level to make it possible to investigate problems
-SSH_OPTIONS+=" -o LogLevel=${SSH_LOG_LEVEL}"
-#######################
-
-# We need root permissions in target container to change ownership of files.
-# If user in that container has access to sudo we use it.
-# Otherwise we use rsync without sudo and if it is possible ownership will be changed.
-RSYNC_COMMAND=$(ssh ${SSH_OPTIONS} ${DST_HOST} \
-            "if hash sudo 2>/dev/null; then echo 'sudo rsync'; else echo 'rsync'; fi")
+source /opt/codenvy-data/conf/rsyncbase.sh
 
 ##### Rsync options #####
 RSYNC_OPTIONS=""
@@ -78,4 +48,4 @@ rsync ${RSYNC_OPTIONS} \
       --rsh="ssh ${SSH_OPTIONS}" \
       --rsync-path="${RSYNC_COMMAND}" \
       --include='.codenvy/' --filter=':- .gitignore' \
-      ${SRC_FOLDER} ${DST_HOST}:${DST_FOLDER}
+      ${SRC_FOLDER} ${HOST}:${DST_FOLDER}

